api/queue/dispatcher.go

```diff
@@ -0,0 +1,131 @@
+package queue
+
+import (
+	"database/sql/driver"
+	"encoding/json"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/gojek/merlin/log"
+	"github.com/jinzhu/gorm"
+)
+
+type Producer interface {
+	EnqueueJob(job *Job) error
+}
+
+type Consumer interface {
+	RegisterJob(jobName string, jobFn func(*Job) error)
+	Start()
+	Stop()
+}
+
+type Dispatcher struct {
+	sync.Mutex
+	db         *gorm.DB
+	workers    []*worker
+	jobFuncMap map[string]JobFn
+	jobChan    chan *Job
+}
+
+type JobFn func(*Job) error
+
+type Config struct {
+	NumWorkers int
+	Db         *gorm.DB
+}
+
+type Status string
+
+type Argument map[string]interface{}
+
+type Job struct {
+	ID        int64     `json:"id" gorm:"primary_key;"`
+	Arguments Argument  `json:"arguments"`
+	Completed bool      `json:"completed"`
+	Name      string    `json:"name"`
+	CreatedAt time.Time `json:"created_at"`
+	UpdatedAt time.Time `json:"updated_at"`
+}
+
+func (a Argument) Value() (driver.Value, error) {
```

Comment:
I think `Arguments` definition should be separated into its own file.

Reply:
will change this